### DISCUSSION TITLE
Quick fixes

### DISCUSSION THREAD

**Discussion by glpnk** [BeardOverflow/msi-ec#130] [Show and tell]
Share the fixes you find here

> [!CAUTION]
> Apply fixes at your own risk, we are not responsible for any damage suffered. 
> Some fixes may don't work for your device or break it.

Template:
```
Device: Tell which device you use

Problem:

Tell here about your issue

Context: [optional]

Tell how thing might work insight

Fix:

Tell how you fixed it
```

**Comment by glpnk** [comment]
### Device:

Modern 14 C5M

### Problem:

Delayed reaction on charger [dis]connection

### Context:

For proper operation of the EC it uses interrupts, which might have different number. Event number declared in ACPI tables, like DSDT and ECDT. In some cases, this number different. Or ECDT table is not configured properly.

### Fix:

This fix required extraction and modification of ECDT table, compiling it and [overriding it by kernel](https://www.kernel.org/doc/Documentation/acpi/initrd_table_override.txt)

#### Steps

1. Follow this [Guide for DSDT patching and overriding](https://gist.github.com/lamperez/d5b385bc0c0c04928211e297a69f32d7), but patch ECDT table instead of DSDT.
2. Bump `Oem Revision` value by 1 (or more) to mark your patch as update, before compiling and packing to CPIO archive.
3. Change `GPE Number` and `Namepath` to values from DSDT for EC section
4. Compile and install according to guide (for systemd-boot bootloader), GRUB2 have easier patching method.
5. Verify (your charger should be detected nearly instant)

**Comment by glpnk** [comment]
### Problem:

Random brightness changes

### Fix:

#### Wayland:
[source](https://askubuntu.com/questions/777754/brightness-randomly-up-and-down-on-msi-laptop/1487963#1487963)

```
[Video Bus Spooky Ghost]
MatchName=*Video Bus*
AttrEventCodeDisable=EV_KEY:0xE0;EV_KEY:0xE1;
```

#### X11: 
[source](https://discussion.fedoraproject.org/t/login-and-screen-brightness-issue-with-fedora-37/79613/8)

Add next values to `/usr/share/X11/xorg.conf.d/10-quirks.conf`
```
Section "InputClass"
 Identifier "Spooky Ghosts"
 MatchProduct "Video Bus"
 Option "Ignore" "on"
EndSection
```

**Comment by glpnk** [comment]
### Problem: 

No Mute-LED trigger

### Fix:

```
echo 'Master Playback Switch' | sudo tee /sys/class/sound/ctl-led/speaker/card1/attach
```

### Credits:

https://github.com/BeardOverflow/msi-ec/issues/156#issue-2501447269

### Fixed by:

https://github.com/alsa-project/alsa-ucm-conf/pull/716 by @Fede2782

**Reply by mutchiko** [reply]
If you get ```No such file or directory``` try changing the card number ```card0``` or ```card2``` for example
```ls /sys/class/sound/ctl-led/speaker``` to check the number of cards.

**Reply by Fede2782** [reply]
I added a fix to Alsa UCM, this should not be required anymore on all distros which will package the new use case manager configs. Mute LED will be automatically linked to Master Playback Switch if msi-ec is installed and is providing the platform::mute LED

**Reply by glpnk** [reply]
Nice. Do you have link for the patch?

**Reply by Fede2782** [reply]
The patch is here and has got cherry picked in the main repo when the issue was closed by the maintainer but has just force pushed things back despite previous commit was already respecting the contibution requirements. I suppose the maintainer is currently working on implementing the auto validation in CI. 

https://github.com/alsa-project/alsa-ucm-conf/pull/716

**Reply by glpnk** [reply]
Thanks

**Comment by glpnk** [comment]
## ECDT patching guide

### Update: use Linux 6.16+

### Device list:

https://github.com/BeardOverflow/msi-ec/discussions/277#discussioncomment-13225756

###  Possible problems:

+ Delayed reaction on charger [dis]connection.

### Debug:

1. Load `EC-SYS` module or `msi-ec` in debug-mode
2. Measure time twice to be sure that time of `sudo` authentication is not included
```
time sudo hexdump -C /sys/kernel/debug/ec/ec0/io
```
__OR__
```
time sudo cat /sys/devices/platform/msi-ec/debug/ec-dump
```
3. Check results (real results may vary):

__OK__
```
real	0m0,056s
user	0m0,003s
sys	0m0,010s
```
__NOT OK__
```
real	0m1,043s
user	0m0,002s
sys	0m0,008s
```

### Context:

For proper operation of the EC it uses interrupts, which might have different number. Event number declared in ACPI tables, like DSDT and ECDT. In some cases, this number different. Or ECDT table is not configured properly.

### Fix:

This fix required extraction and modification of ECDT table, compiling it and [overriding it by kernel](https://www.kernel.org/doc/Documentation/acpi/initrd_table_override.txt)

Older guide https://github.com/BeardOverflow/msi-ec/discussions/130#discussioncomment-9694621

### Steps:

1. Get `iasl`
+ `acpica` - Arch Linux
+ `acpica-tools` - Debian, Ubuntu

2. Get ECDT
```
sudo cat /sys/firmware/acpi/tables/ECDT > ecdt-mod.dat
```

5. Decompile
```
iasl -d ecdt-mod.dat
```

6. Fix `GPE`, `Namepath` in `ecdt-mod.dsl` file: (__example, real values could be found in DSDT/SSDT__)
```patch
- [03Ch 0060 004h]                         UID : 00000000
- [040h 0064 001h]                  GPE Number : 00
- [041h 0065 001h]                    Namepath : ""

+ [03Ch 0060 004h]                         UID : 00000000
+ [040h 0064 001h]                  GPE Number : 07
+ [041h 0065 001h]                    Namepath : "\\_SB.PCI0.SBRG.EC"
```

7. Bump `Oem Revision`
```patch
- [018h 0024 004h]                Oem Revision : 01072009

+ [018h 0024 004h]                Oem Revision : 01072010
```

8. Compile
```
iasl -tc ecdt-mod.dsl
```

9. Use `ecdt-mod.aml` file as:
+ Initramfs patch, similar to [DSDT fix in archwiki](https://wiki.archlinux.org/title/DSDT)
  + [mkinitcpio](https://wiki.archlinux.org/title/DSDT#Using_mkinitcpio's_acpi_override_hook)
  + [dracut](https://wiki.archlinux.org/title/DSDT#Using_the_AML_with_dracut)
+ payload for bootloader
  + [GRUB2](https://wiki.archlinux.org/title/DSDT#Using_the_AML_with_GRUB)
  + [SystemD-Boot](https://gist.github.com/lamperez/d5b385bc0c0c04928211e297a69f32d7#install-the-tables-systemd-boot)
10. Reboot
11. Test EC reading time
